### PR TITLE
Extract MetricMap

### PR DESCRIPTION
**Pull request by tiedotguy**
This is a "small" isolated commit to extract the MetricMap Receive logic out from the MetricAggregator.

It's mostly noise and moving things around unfortunately, but better the noise be in its own PR than in what follows.

This will be re-pointed to master once http-transport is merged.

### CODE REVIEW

File: metric_map.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	case COUNTER:
+		mm.receiveCounter(m, tagsKey, nowNano)
+	case GAUGE:
+		mm.receiveGauge(m, tagsKey, nowNano)
+	case TIMER:
+		mm.receiveTimer(m, tagsKey, nowNano)
+	case SET:
+		mm.receiveSet(m, tagsKey, nowNano)
+	default:
+		logrus.StandardLogger().Errorf("Unknown metric type %s for %s", m.Type, m.Name)
```
Comment (aelse) [inline metric_map.go]: Is there sanitisation on m.Type before we reach this point? I don't like the idea of emitting error logs on bad user input (bad statsd clients / unsupported types) but if we should've caught it before here that's another matter.

File: metric_map.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+	}
+	m.Done()
+}
+
+func (mm *MetricMap) Merge(mmFrom *MetricMap) {
+	mmFrom.Counters.Each(func(metricName string, tagsKey string, counterFrom Counter) {
+		v, ok := mm.Counters[metricName]
+		if ok {
```
Comment (aelse) [inline metric_map.go]: consider handling the !ok case first and returning early to avoid all these nested ifs

Comment (aelse) [reply metric_map.go]: and for the functions following this one also

**Comment by tiedotguy** [comment]
I agree there's issues (I've wanted to redo the logic on those `receive*` calls for a long time), but this is a "lift and shift" refactor only, none of this is new code, just new locations for it.

For the sanitization generally, it only happens in tests.  It's one of those "this should panic to make the test fail, but we won't *just in case* it's not a test" situations.

I'll raise an issue for refactoring the `receive*` calls later.

**Pull request merged**